### ISSUE TITLE
Kotlin coroutine missing trace

### ISSUE THREAD

**Issue by asarkar** [open-telemetry/opentelemetry-java-instrumentation#1550]
**Describe the bug**
Code snippet from a gRPC `ClientInterceptor`:
```
private fun getOrCreateUuid(userId: String): String {
    ...
    CoroutineScope(AppGlobal.IODispatcher + AppGlobal.ExceptionHandler).launch {
        uuidRepository.save(uuid)
        LOG.debug("Saved user id: {} in the DB", userId)
    }
}
```
The log statement is missing trace and span, but log statements outside the `launch` coroutine have trace and span. 

**Steps to reproduce**
Launch a coroutine from inside a gRPC `ClientInterceptor`.

**What did you expect to see?**
Trace and span in the log statements.

**What did you see instead?**
Trace and span missing in the log statements.

**What version are you using?**
v0.9.0

**Environment**
N/A.

**Additional context**
N/A.

**Comment by anuraaga** [comment]
Hi @asarkar - will https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/1617 give you the helper you need to solve this? I believe this would work

```
CoroutineScope(AppGlobal.IODispatcher + Context.current().asContextElement()).launch {
  ..
}
```

**Comment by asarkar** [comment]
@anuraaga Where is the method `asContextElement()` coming from, I'm not seeing it. Is it an extension function from a library not part of coroutine core?

**Comment by anuraaga** [comment]
@asarkar It's this extension method here - https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/1617/files#diff-10f00e16c51ae604be10e6a3af1fb116754fc4ad2f9f322528424afe45d3e744

**Comment by asarkar** [comment]
I don't understand, if it's a pending PR, how's it supposed to help me now? Don't I need to wait for it to be merged?

**Comment by anuraaga** [comment]
@asarkar Mostly so you can check it to see if anything looks really bad with it, or maybe copy-paste it to try. But otherwise, we're hoping it will help after it's merged :)

**Comment by anuraaga** [comment]
@asarkar The latest snapshots of the SDK include a new artifact, `opentelemetry-extension-kotlin` which includes the helper method.

https://github.com/open-telemetry/opentelemetry-java/blob/master/extensions/kotlin/src/main/kotlin/io/opentelemetry/extension/kotlin/ContextExtensions.kt

**Comment by asarkar** [comment]
I’ll try it out, and let you know. If I understood correctly, you’re suggesting that the user add an element to the coroutine context whenever they launch a new coroutine. That’d be tedious and error prone, isn’t it? Is it possible to install the trace in the parent coroutine and the have it propagated transparently to the children?

**Comment by anuraaga** [comment]
@asarkar Ah just to confirm, are you using the agent or just the library instrumentation? 

Basically, the context needs to be added to the coroutine context just once when launching the top-level, and it would propagate to any children coroutines. Our library instrumentation can't do this since they're not implemented as coroutines, we do need the user to setup the context once when they first launch a coroutine. With #1618, the agent will intercept calls to `launch` to do this automatically for agent users though.

**Comment by asarkar** [comment]
I'll be using the library.

**Comment by asarkar** [comment]
@anuraaga I tried with the following versions:
Group `io.opentelemetry.instrumentation` - v0.10.1
Group `io.opentelemetry` - v0.12.0-SNAPSHOT

but:
```
java.lang.NoSuchMethodError: 'io.opentelemetry.api.trace.Span$Builder io.opentelemetry.api.trace.Tracer.spanBuilder(java.lang.String)'
	at io.opentelemetry.instrumentation.grpc.v1_5.client.GrpcClientTracer.startSpan(GrpcClientTracer.java:26)
	at io.opentelemetry.instrumentation.grpc.v1_5.client.TracingClientInterceptor.interceptCall(TracingClientInterceptor.java:57)
```

**Comment by anuraaga** [comment]
Yeah until 1.0, we are not creating compatibility bridges like we will going post-1.0, sorry for the trouble. In the meantime, are you able to use `0.11-0-SNAPSHOT` of instrumentation along with `0.11.0` of the SDK?

**Comment by asarkar** [comment]
@anuraaga Using the versions you suggested, I gathered the following logs from a test execution (sanitized to protect IP). There seems to be more than one discrepancies.

```
{"@timestamp":"2020-12-08T21:17:16.390Z","severity":"INFO","trace":"","span":"","thread":"Test worker @coroutine#2","class":"MetadataInjectingClientInterceptor","rest":"Adding request header: ***","exception":""}
{"@timestamp":"2020-12-08T21:17:16.439Z","severity":"DEBUG","trace":"fbe59ddaeeb27d04dd243adbc0838a0c","span":"a9e2175520bc4444","thread":"grpc-default-executor-0","class":"MetadataTransferringServerInterceptor","rest":"Copying headers to context ","exception":""}
{"@timestamp":"2020-12-08T21:17:16.491Z","severity":"INFO","trace":"","span":"","thread":"DefaultDispatcher-worker-1 @coroutine#5","class":"MetadataInjectingClientInterceptor","rest":"Adding request header: ***","exception":""}
{"@timestamp":"2020-12-08T21:17:16.503Z","severity":"DEBUG","trace":"b03b1c53f0d98c4042e290b24e81a02b","span":"260a5cc67b051040","thread":"DefaultDispatcher-worker-1 @coroutine#6","class":"AuthInterceptor","rest":"User id: ***","exception":""}
{"@timestamp":"2020-12-08T21:17:16.567Z","severity":"INFO","trace":"b03b1c53f0d98c4042e290b24e81a02b","span":"260a5cc67b051040","thread":"DefaultDispatcher-worker-1 @coroutine#6","class":"HttpClient","rest":"REQUEST: http://localhost:7777/v2/authorize ","exception":""}
// More HttpClient logs
{"@timestamp":"2020-12-08T21:17:16.726Z","severity":"INFO","trace":"","span":"","thread":"qtp343044267-33","class":"ROOT","rest":"RequestHandlerClass from context returned com.github.tomakehurst.wiremock.http.StubRequestHandler. Normalized mapped under returned 'null' ","exception":""}
{"@timestamp":"2020-12-08T21:17:16.856Z","severity":"DEBUG","trace":"b03b1c53f0d98c4042e290b24e81a02b","span":"260a5cc67b051040","thread":"DefaultDispatcher-worker-1 @coroutine#6","class":"ClientRegistrar","rest":"Registration successful","exception":""}
{"@timestamp":"2020-12-08T21:17:16.861Z","severity":"DEBUG","trace":"b03b1c53f0d98c4042e290b24e81a02b","span":"260a5cc67b051040","thread":"DefaultDispatcher-worker-1 @coroutine#5","class":"AuthInterceptor","rest":"User id: ***","exception":""}
{"@timestamp":"2020-12-08T21:17:16.862Z","severity":"INFO","trace":"b03b1c53f0d98c4042e290b24e81a02b","span":"260a5cc67b051040","thread":"kt-io-1 @coroutine#22","class":"AuthInterceptor","rest":"Saved user id: *** in the DB ","exception":""}
{"@timestamp":"2020-12-08T21:17:16.924Z","severity":"DEBUG","trace":"b03b1c53f0d98c4042e290b24e81a02b","span":"260a5cc67b051040","thread":"DefaultDispatcher-worker-1 @SendMessage worker for ***","class":"LoggingInterceptor","rest":"Sending request","exception":""}
{"@timestamp":"2020-12-08T21:17:17.448Z","severity":"DEBUG","trace":"","span":"","thread":"grpc-default-executor-0","class":"LoggingInterceptor","rest":"Received response","exception":""}
{"@timestamp":"2020-12-08T21:17:17.587Z","severity":"INFO","trace":"","span":"","thread":"DefaultDispatcher-worker-1 @coroutine#5","class":"audio","rest":"Done sending data ","exception":""}
{"@timestamp":"2020-12-08T21:17:17.592Z","severity":"INFO","trace":"","span":"","thread":"DefaultDispatcher-worker-1 @coroutine#5","class":"audio","rest":"Done sending silence ","exception":""}
{"@timestamp":"2020-12-08T21:17:17.948Z","severity":"DEBUG","trace":"","span":"","thread":"grpc-default-executor-2","class":"LoggingInterceptor","rest":"Received response","exception":""}
```

First of all, we need to establish the call flow.

MetadataInjectingClientInterceptor - gRPC client interceptor, only used for testing.
MetadataTransferringServerInterceptor - gRPC server interceptor
AuthInterceptor - gRPC client interceptor
HttpClient - Ktor Http client
ClientRegistrar - Class that uses Ktor Http client
LoggingInterceptor  - gRPC client interceptor

MetadataInjectingClientInterceptor -> MetadataTransferringServerInterceptor -> Unnamed gRPC service -> AuthInterceptor/HttpClient/ClientRegistrar/LoggingInterceptor  -> gRPC client -> LoggingInterceptor

1. > the context needs to be added to the coroutine context just once when launching the top-level

This didn't work. The context needed to be added to the nearest coroutine context in scope.

2. The span changes between `MetadataTransferringServerInterceptor` (gRPC  and `AuthInterceptor`. Note that the gRPC service starts a new context:
```
withContext(AppGlobal.IODispatcher + AppGlobal.ExceptionHandler) {
  launch {...}
}
```

3. `HttpClient` logs should have a different span, since it's a new HTTP call.

4. `LoggingInterceptor` "Received response" is missing trace/span. However, `LoggingInterceptor` "Sending request" has trace/span, although it's the same class (but different threads).

5. `audio` (top-level coroutines) logs are missing trace/span.

**Comment by anuraaga** [comment]
@asarkar Are all the classes implemented in kotlin? Are you using `grpc-kotlin`? 

https://github.com/grpc/grpc-kotlin 

They have an interceptor specially for adding context elements, which I think you would want to use if you need it to propagate among all those components

https://javadocs.dev/io.grpc/grpc-kotlin-stub/1.0.0/index.html

**Comment by asarkar** [comment]
> Are all the classes implemented in kotlin? Are you using grpc-kotlin?

Yes, and yes.

> They have an interceptor

Are you referring to the [CoroutineContextServerInterceptor](https://javadocs.dev/io.grpc/grpc-kotlin-stub/1.0.0/io/grpc/kotlin/CoroutineContextServerInterceptor.html)? The trace goes missing when a new coroutine is launched from a client interceptor or from a top-level coroutine; it's not clear to me how the `CoroutineContextServerInterceptor` can be used to fix this issue.

**Comment by anuraaga** [comment]
@asarkar Ah yeah that interceptor. If that interceptor is installed first and returns e.g. `Context.current().asContextElement()`, that should cause Kotlin's context propagation mechanism to carry it through coroutines. By default, coroutines inherit the context of their parent. Does it not work? If you can give some example code we can try to figure out what's up and add some unit tests here too

**Comment by asarkar** [comment]
@anuraaga 
Using the following, and configuring logback as instructed [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/v0.16.x/instrumentation/logback/logback-1.0/library), I can see gRPC trace in the logs.
```
io.opentelemetry.instrumentation:opentelemetry-grpc-1.5:0.9.0
io.opentelemetry.instrumentation:opentelemetry-logback-1.0.0:0.9.0
io.opentelemetry:opentelemetry-sdk-tracing:0.9.1
```

However, upgrade to `0.16.0`, and the trace disappears.
```
io.opentelemetry.instrumentation:opentelemetry-grpc-1.5:0.16.0
io.opentelemetry.instrumentation:opentelemetry-logback-1.0:0.16.0
io.opentelemetry:opentelemetry-sdk-trace:0.16.0
```

**Comment by anuraaga** [comment]
Hi @asarkar - can you confirm with 0.17.0 to make sure we're on the same page? Mostly because in 0.17.0 the MDC keys were renamed to e.g., `trace_id` so it'll be hard to share example code if not synced

https://github.com/open-telemetry/opentelemetry-java-instrumentation/releases/tag/v0.17.0

**Comment by asarkar** [comment]
I noticed the rename. Also, last I checked, both groups hadn't released all libraries with 0.17.0 version. However, I'll try and report back.

**Comment by asarkar** [comment]
@anuraaga I can confirm that using `0.17.0-alpha` and the [Logback Integration](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/logback/logback-1.0/library) has the same issue.
```
io.opentelemetry.instrumentation:opentelemetry-instrumentation-bom-alpha:0.17.0-alpha
io.opentelemetry.instrumentation:opentelemetry-instrumentation-api
io.opentelemetry.instrumentation:opentelemetry-grpc-1.5
io.opentelemetry.instrumentation:opentelemetry-logback-1.0
io.opentelemetry:opentelemetry-sdk-trace
```

Am I missing a runtime dependency, or is this something to do with gRPC and Kotlin? Like I said, v0.9.0 works fine.

**Comment by asarkar** [comment]
Here's a fully reproducible application. Simply run `./gradlew test` and watch it fail.

[otel-grpc-tracing 2.zip](https://github.com/open-telemetry/opentelemetry-java-instrumentation/files/6088437/otel-grpc-tracing.2.zip)

**Comment by anuraaga** [comment]
Thanks for the app @asarkar. Since that time, the behavior of the SDK changed to default to being completely no-op - you have to configure the SDK yourself once to set it up. I changed your app to have an `implementation` dependency on `io.opentelemetry:opentelemetry-sdk` and modified `Application` to be

```java
@SpringBootApplication
class Application {
    private val LOG: Logger = LoggerFactory.getLogger(Application::class.java)

    @Bean
    fun openTelemetry(): OpenTelemetry {
        return OpenTelemetrySdk.builder()
            .buildAndRegisterGlobal()
    }

    @Bean
    @GrpcGlobalServerInterceptor
    // Run after MetricCollectingServerInterceptor
    @Order(InterceptorOrder.ORDER_TRACING_METRICS + 1)
    fun tracingServerInterceptor(openTelemetry: OpenTelemetry): ServerInterceptor {
        LOG.info("gRPC server tracing is enabled")
        return TracingServerInterceptor.newInterceptor(openTelemetry.getTracer("grpc"))
    }

    @Bean
    @GrpcGlobalClientInterceptor
    // Run after MetricCollectingClientInterceptor
    @Order(InterceptorOrder.ORDER_TRACING_METRICS + 1)
    fun tracingClientInterceptor(openTelemetry: OpenTelemetry): ClientInterceptor {
        LOG.info("gRPC client tracing is enabled")
        return TracingClientInterceptor.newInterceptor(openTelemetry.getTracer("grpc"))
    }
}
```

and it seems to work ok.

FYI, in the next release, the entrypoint to gRPC will change to `GrpcTracing.create(openTelemetry).newServerInterceptor()` / `GrpcTracing.create(openTelemetry).newClientInterceptor()`

**Comment by asarkar** [comment]
@anuraaga Thank you for looking into this. It'd have saved us, and future users, a _lot_ of time if something like the following was logged.
```
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
```

**Comment by asarkar** [comment]
Now I'm getting this; I think the next version you mentioned should avoid it. How do I get it?
```
Caused by: java.lang.IllegalStateException: GlobalOpenTelemetry.set has already been called. GlobalOpenTelemetry.set must be called only once before any calls to GlobalOpenTelemetry.get. If you are using the OpenTelemetrySdk, use OpenTelemetrySdkBuilder.buildAndRegisterGlobal instead. Previous invocation set to cause of this exception.
	at io.opentelemetry.api.GlobalOpenTelemetry.set(GlobalOpenTelemetry.java:83)
	at io.opentelemetry.sdk.OpenTelemetrySdkBuilder.buildAndRegisterGlobal(OpenTelemetrySdkBuilder.java:62)
```

**Comment by anuraaga** [comment]
>  It'd have saved us, and future users, a lot of time if something like the following was logged.

Yeah I had a PR out for that a while ago, during some other work it seemed like it might not be necessary but with this I realized we do still need it so resurrected it.
https://github.com/open-telemetry/opentelemetry-java/pull/2396

> Now I'm getting this if there are two tests in the same class; I think the next version you mentioned should avoid it. How do I get it?

Even in the current code you can replace `buildAndRegisterGlobal` with just `build` if you'd like. Caveat is that while currently no library instrumentation requires the use of the global, in the future some niche ones may (mysql-connector-java probably will at some point). For tests you can call `GlobalOpenTelemetry.resetForTest()` at the beginning if needed.

We also provide a junit5 extension, but it may not work for an actual spring integration test easily - if we can find patterns for using that along with spring integration tests though it will be great.

**Comment by asarkar** [comment]
I tried with just `build` and it didn't work. See for yourself in the sample app. The problem seems to be the gRPC tracers don't expose the super constructor that accepts an OpenTelemetry instance.

**Comment by anuraaga** [comment]
Ah ok - you can try the `0.18.0-SNAPSHOT` builds if you'd like

https://oss.sonatype.org/content/repositories/snapshots/io/opentelemetry/instrumentation/opentelemetry-grpc-1.5/0.18.0-alpha-SNAPSHOT/

though we should be releasing `1.0.0-alpha` by tomorrow so it could be safer to just wait for that.

**Comment by asarkar** [comment]
@anuraaga Using `0.18.0-alpha-SNAPSHOT` seems to work for the sample app, but not for the real one that uses grpc-kotlin. I'm looking....

Also, FYI, the following warning is shown by IntelliJ at `GrpcTracing.create(openTelemetry).newServerInterceptor()` call site. This is because the `GrpcTracing` constructor is package scoped. It doesn't seem to affect anything, but you may want to change it to `public` if possible. A similar warning is shown for the client interceptor.
```
Type TracingServerInterceptor! is inaccessible in this context due to: public/*package*/ final class TracingServerInterceptor : ServerInterceptor defined in io.opentelemetry.instrumentation.grpc.v1_5 in file TracingServerInterceptor.class
```

**Comment by anuraaga** [comment]
> Also, FYI, the following warning is shown by IntelliJ at GrpcTracing.create(openTelemetry).newServerInterceptor() call site.

Oops, thanks! #2510 should fix that.

**Comment by asarkar** [comment]
@anuraaga After doing what you had suggested in [this comment](https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/1550#issuecomment-743706550), I can now see server side trace, but client side trace is still absent. If you see my [earlier comment](https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/1550#issuecomment-742001851), the `LoggingInterceptor` and the `AuthInterceptor` are client side, and they used to show trace. Now, they don't. grpc-kotlin doesn't provide a client interceptor like the server one, what do I do to restore client side trace?

**Comment by anuraaga** [comment]
@asarkar So the client interceptors could see the trace with 0.9.0 but not with the new version? I can't think of any change between the versions that would have caused that issue, we have this unit test which is verifying there's a span in the client interceptor

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/grpc-1.5/testing/src/main/groovy/io/opentelemetry/instrumentation/grpc/v1_5/AbstractGrpcTest.groovy#L348

Are you starting the coroutine in `ClientCall.start`? If you're able create a simple reproduction than I can take a look at it.

**Comment by asarkar** [comment]
@anuraaga Something very weird happened here, I thought I responded to your comment, but I don't see it. Perhaps I forgot to press the "Comment" button :(

Anyway, the gRPC server calls a gRPC client in a new Coroutine context.
```
withContext(AppGlobal.IODispatcher + AppGlobal.ExceptionHandler) {
	launch { // gRPC client call }
}
```
I added a `ServerInterceptor` as you suggested, and that seems to propagate the OTEL context. Note that _this wasn't required with v0.9.0_. Also note that the call is made using a custom dispatcher.

The client `LoggingInterceptor` is more or less like this:

```
class LoggingInterceptor : ClientInterceptor, Ordered {
    private val LOG: Logger = LoggerFactory.getLogger(LoggingInterceptor::class.java)
    private val PRINTER = TextFormat.printer().escapingNonAscii(false)

    override fun <ReqT, RespT> interceptCall(
        method: MethodDescriptor<ReqT, RespT>,
        callOptions: CallOptions,
        next: Channel
    ): ClientCall<ReqT, RespT> {
        return object : SimpleForwardingClientCall<ReqT, RespT>(
            next.newCall(method, callOptions)
        ) {
            override fun sendMessage(req: ReqT) {
                LOG.debug("Sending request: {}", PRINTER.printToString(req))

                super.sendMessage(req)
            }

            override fun start(responseListener: Listener<RespT>, headers: GRPCMetadata) {
                val listener: Listener<RespT> =
                    object : ForwardingClientCallListener.SimpleForwardingClientCallListener<RespT>(responseListener) {
                        override fun onMessage(res: RespT) {
                            LOG.debug("Received response: {}", PRINTER.printToString(res))

                            super.onMessage(res)
                        }
                    }
                super.start(listener, headers)
            }
        }
    }

    override fun getOrder() = ORDER

    companion object {
        const val ORDER = AuthInterceptor.ORDER + 1
    }
}
```

The `sendMessage` executes on the custom dispatcher set up by the server, but the `onMessage` executes on gRPC default thread pool, and is missing the trace. _It used to work with v0.9.0_.

**Comment by asarkar** [comment]
Here's an sample app with a test demonstrating the issue.
[otel-grpc-tracing.zip](https://github.com/open-telemetry/opentelemetry-java-instrumentation/files/6112363/otel-grpc-tracing.zip)

**Comment by asarkar** [comment]
@anuraaga Have you had a chance to look?

**Comment by anuraaga** [comment]
I have a feeling this is because of the switch off of using grpc context directly for OTel. So it's time to implement a proper context bridge it'll take some time.

**Comment by asarkar** [comment]
Also,
```
CoroutineScope(AppGlobal.IODispatcher + Context.current().asContextElement()).launch { ... }
```

User code shouldn't need to do this. I believe OTEL should be able to hook into coroutine creation and suspension to propagate the context. [kotlinx-coroutines-debug](https://github.com/Kotlin/kotlinx.coroutines/tree/master/kotlinx-coroutines-debug) does exactly that, although I've not looked into how.

**Comment by anuraaga** [comment]
kotlinx-coroutines-debug is a Java agent - our agent also similarly applies that pattern automatically so users don't have to.

For library users without the agent there isn't a way to avoid the code, though for grpc-kotlin the context interceptor should be making it relatively simple, coroutines should inherit the parent context set by the interceptor automatically I think.

**Comment by asarkar** [comment]
Sadly, I can confirm that without the above, user launched coroutines lose the trace.
There are two distinct issues here:

1. gRPC trace propagation.
2. Kotlin coroutine trace propagation.

For many users, me included, both coexist. But I'm happy to create another ticket for you specifically for gRPC tracing (or you can create yourself), so that work on each can be done independently without holding up the other.

**Comment by anuraaga** [comment]
@asarkar The kotlin aspect is general usage of kotlin coroutines, there isn't really anything otel-specific (though otel is generally going to be the main consumer of context in practice). We unfortunately don't have any Kotlin experts here, so perhaps there is a kotlin repo (grpc-kotlin?) where the support can be more targeted.

As a kotlin amateur, I can only infer from docs, but I think you're not supposed to use `CoroutineScope(...)` - that creates a new scope so naturally loses the context, you should just have one scope per request. You can create a new child coroutine within the same scope, which does inherit context, using just `launch(IO)`

https://kotlinlang.org/docs/coroutine-context-and-dispatchers.html#children-of-a-coroutine

I've updated my SDK unit test which was broken in a couple of ways to demonstrate the pattern.

https://github.com/open-telemetry/opentelemetry-java/pull/3012

**Comment by asarkar** [comment]
I’m certainly no Kotlin ninja, but I’m not an amateur either, so if there’s something I can help with, let me know. We can, of course, raise some Kotlin specific issues to grpc-Kotlin, but tracing gRPC calls is not one of their requirements, so they may not be very forthcoming. Also, last I checked, they had a very small team working on that project.

**Comment by anuraaga** [comment]
@asarkar So do you expect `CoroutineScope(AppGlobal.IODispatcher).launch` to automatically have context propagated? Based on the docs, this seems to not be the correct way to launch a child coroutine, `launch(AppGlobal.IODispatcher)` within a coroutine would work though. If that's correct, then there doesn't seem to be anything left kotlin-specific here

**Comment by asarkar** [comment]
> kotlinx-coroutines-debug is a Java agent 

This isn't true, at least not entirely. I've only used it as a library, where the hooks are installed by making explicit calls from my code. In the case of gRPC, I assume an interceptor would install these hooks.

> this seems to not be the correct way to launch a child coroutine

[`GlobalScope` is discouraged](https://elizarov.medium.com/the-reason-to-avoid-globalscope-835337445abc), but unless I'm missing something, there's nothing wrong with creating a new scope. In fact, coroutine exception handling behavior is very different between the two ways (`scope` vs `launch`). I've done a lot of [tests](https://github.com/asarkar/kotlin/blob/master/coroutines-learning/src/test/kotlin/com/asarkar/kotlin/TestCoroutinesExceptionHandling.kt) about that. Refer to the articles in the README of that project for details if you're interested.

**Comment by anuraaga** [comment]
> there's nothing wrong with creating a new scope

There's nothing wrong with invoking any sort of code naturally :) But if a Scope in kotlin means a new root context, which does seem to be the case, then having context not propagated automatically seems to be correct behavior.

In those unit tests, I don't see any of the test methods creating a new scope, as your sample code `CoroutineScope(AppGlobal.IODispatcher + Context.current().asContextElement()).launch { ... }` does, they always use instance methods on the `scope` you pass in, except for this one using `GlobalScope`

https://github.com/asarkar/kotlin/blob/master/coroutines-learning/src/test/kotlin/com/asarkar/kotlin/TestCoroutinesExceptionHandling.kt#L397

and if context isn't propagated to that method, it seems correct, the docs very explicitly say `However, when GlobalScope is used to launch a coroutine, there is no parent for the job of the new coroutine. It is therefore not tied to the scope it was launched from and operates independently.`. I have to check the unit tests for our javaagent since I have a feeling we may actually be automatically propagating into unrelated Scopes and that would actually be a bug.

**Comment by asarkar** [comment]
> I don't see any of the test methods creating a new scope

All `ParameterizedTest` methods with [scopeProvider](https://github.com/asarkar/kotlin/blob/master/coroutines-learning/src/test/kotlin/com/asarkar/kotlin/TestCoroutinesExceptionHandling.kt#L404) do.

> a Scope in kotlin means a new root context

It doesn't. From [coroutineScope Kdoc](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/coroutine-scope.html):

> Creates a CoroutineScope and calls the specified suspend block with this scope. The provided scope inherits its coroutineContext from the outer scope, but overrides the context’s Job.

So, if the outer scope had a OTEL context as a context element, the inner should too. You see why I think losing the scope in the user-launched coroutine is a bug? I guess I can debug this myself, but I want to convince you that a major problem exists.

[This article](https://elizarov.medium.com/coroutine-context-and-scope-c8b255d59055) from the lead developer of Kotlin coroutine explains the concepts better than I can.

**Comment by anuraaga** [comment]
@asarkar Did you find this method invocation `coroutineScope` to also not work? Your example code that you said fails is a constructor invocation `CoroutineScope`, which is not intended to create a "child scope" as far as I can tell.

Added `coroutineScope` and `CoroutineScope` to https://github.com/open-telemetry/opentelemetry-java/pull/3012 to illustrate

**Comment by asarkar** [comment]
`coroutineScope` (small 'c') is a suspending function that must be called from within another coroutine, whereas `CoroutineScope` (capital 'C) is not a suspending function and can be called from within regular function. Depending on the call site, you may not be able to use them alternatively.

You have pointed out an important point though, I'll verify if I can use `coroutineScope` (small 'c') and report back what I find. There's a [SO post](https://stackoverflow.com/q/59368838/839733) about the differences, but it's based on Android, not JVM, so not directly comparable.

**Comment by anuraaga** [comment]
> is not a suspending function and can be called from within regular function. Depending on the call site, you may not be able to use them alternatively.

Yup - when going from Java into the Kotlin coroutine world you'll need to create the root coroutine scope, and when creating the scope, you have to define its context - this is just the nature of the JVM not having a standard concept of Context (in a couple of decades I guess fibers will have introduced something though). We provide the `asContextElement` helper for that, and the javaagent does this automatically because the javaagent is able to do magic that otherwise requires users to tell Kotlin how to do it. If Kotlin provided some static method, `SetCoroutineInitializer` which would be called anytime a new coroutine scope is created and allowed for adding context elements, then that would be the easiest way of doing it, but I guess that doesn't exist, so we need to call `asContextElement` whenever creating a new root.

Think of `asContextElement` as the mechanism to propagate from Java into a coroutine, but it's not required for suspending function into other suspending functions. You may find yourself in Java again and wanting to launch another coroutine though - technically you probably want to propagate `coroutineScope` itself through the Java to that point, but it might be impractical and you just go with a new root Coroutine. We're back to the case of Java calling into coroutine again, and need to call `asContextElement` again - I think this is the best we can do lacking something like that static global CoroutineInitializer type of concept.

**Comment by anuraaga** [comment]
> This isn't true, at least not entirely. I've only used it as a library, where the hooks are installed by making explicit calls from my code. In the case of gRPC, I assume an interceptor would install these hooks.

By the way, I use the term agent loosely as something that uses bytebuddy to intrusively rewrite classes in the JVM - both this javaagent and the kotlin debug mechanism are based on the same premise. For this repo, we currently expect people that are ok with such intrusion to just be using our agent with the startup flag instead of applying through runtime hooks - we could consider providing a runtime version of some parts like this context propagation if it really helps, but I'd rather see Kotlin provide a hook instead of us forcing our ways inside for library users.

**Comment by anuraaga** [comment]
@asarkar I found https://github.com/grpc/grpc-kotlin/issues/223 which would probably help since it feels like the kotlin -> java -> back to kotlin aspect introduced by your interceptor is the brunt of the trickiness. Perhaps you can push on it? :)

**Comment by asarkar** [comment]
https://github.com/grpc/grpc-kotlin/issues/223 is on the server side, whereas I do auth on the client side while calling a third-party service (my server doesn't need auth, it's an internal service). I don't see the similarities of the top of my head, but I'll double check in the morning when I'm not half asleep.

**Comment by asarkar** [comment]
Created https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/2564 to trace the gRPC trace issue. Closing this ticket as done, seems the Kotlin extension is working as designed. `CoroutineScope` creates a new scope, so the trace has to be propagated manually.
